wasm/binary/decoder.go

```diff
@@ -0,0 +1,124 @@
+package binary
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+
+	"github.com/tetratelabs/wazero/wasm"
+	"github.com/tetratelabs/wazero/wasm/leb128"
+)
+
+// TODO: maybe io.ByteReader
+type reader struct {
+	binary []byte
+	read   int
+	buffer *bytes.Buffer
+}
+
+func (r *reader) Read(p []byte) (n int, err error) {
+	n, err = r.buffer.Read(p)
+	r.read += n
+	return
+}
+
+// DecodeModule implements wasm.DecodeModule for the WebAssembly 1.0 (MVP) Binary Format
+// See https://www.w3.org/TR/wasm-core-1/#binary-format%E2%91%A0
+func DecodeModule(binary []byte) (*wasm.Module, error) {
+	r := &reader{binary: binary, buffer: bytes.NewBuffer(binary)}
+
+	// Magic number.
+	buf := make([]byte, 4)
+	if _, err := io.ReadFull(r, buf); err != nil || !bytes.Equal(buf, magic) {
+		return nil, ErrInvalidMagicNumber
+	}
+
+	// Version.
+	if _, err := io.ReadFull(r, buf); err != nil || !bytes.Equal(buf, version) {
+		return nil, ErrInvalidVersion
+	}
+
+	m := &wasm.Module{}
+	for {
+		sectionID := make([]byte, 1)
+		if _, err := io.ReadFull(r, sectionID); err == io.EOF {
+			break
+		} else if err != nil {
+			return nil, fmt.Errorf("read section id: %w", err)
+		}
+
+		sectionSize, _, err := leb128.DecodeUint32(r)
+		if err != nil {
+			return nil, fmt.Errorf("get size of section for id=%d: %v", sectionID[0], err)
+		}
+
+		sectionContentStart := r.read
+		switch sectionID[0] {
+		case SectionIDCustom:
+			// First, validate the section and determine if the section for this name has already been set
+			name, dataSize, decodeErr := decodeCustomSectionNameAndDataSize(r, sectionSize)
```

Comment:
I left this section here as it is the only one special (this avoids having to pass module to each of the type decoders, as it is only needed here)